lib/Transforms/LLZKPCLLoweringPass.cpp

```diff
@@ -0,0 +1,427 @@
+          auto eq = lowerBinaryLike<CmpOp, pcl::CmpEqOp>(b, m, llzkToPcl);
+          if (failed(eq)) {
+            res = eq;
+            break;
+          }
+          auto eqRes = lookup(m.getResult(), llzkToPcl, m.getOperation());
+          if (failed(eqRes)) {
+            res = failure();
+            break;
+          }
+          auto loc = m.getLoc();
+          auto neg = b.create<pcl::NegOp>(loc, *eqRes);
+          b.create<pcl::AssertOp>(loc, neg.getRes());
+          break;
+        }
+        case FeltCmpPredicate::LT:
+          res = lowerBinaryLike<CmpOp, pcl::CmpLtOp>(b, m, llzkToPcl);
+          break;
+        case FeltCmpPredicate::LE:
+          res = lowerBinaryLike<CmpOp, pcl::CmpLeOp>(b, m, llzkToPcl);
+          break;
+        case FeltCmpPredicate::GT:
+          res = lowerBinaryLike<CmpOp, pcl::CmpGtOp>(b, m, llzkToPcl);
+          break;
+        case FeltCmpPredicate::GE:
+          res = lowerBinaryLike<CmpOp, pcl::CmpGeOp>(b, m, llzkToPcl);
+          break;
+        }
+      })
+          .Case<EmitEqualityOp>([&b, &llzkToPcl, &res](EmitEqualityOp m) {
+        auto lhs = lookup(m.getLhs(), llzkToPcl, m.getOperation());
+        auto rhs = lookup(m.getRhs(), llzkToPcl, m.getOperation());
+        if (failed(lhs) || failed(rhs)) {
+          res = failure();
+          return;
+        }
+
+        Value lhsVal = *lhs, rhsVal = *rhs;
+        auto loc = m.getLoc();
+        if (failed(emitAssertEqOptimized(b, loc, lhsVal, rhsVal))) {
+          res = failure();
+          return;
+        }
+      })
+          .Case<FieldReadOp>([&field2pclvar, &llzkToPcl, &srcFunc](FieldReadOp read) {
+        // At this point every field in the struct should have a var associated with it
+        // so we should simply retrieve the var associated with the field.
+        assert(read.getComponent() == srcFunc.getArguments()[0]);
+        if (auto it = field2pclvar.find(read.getFieldName()); it != field2pclvar.end()) {
+          rememberResult(read.getResult(), it->getSecond(), llzkToPcl);
+        } else {
+          llvm_unreachable("Every field should have been mapped to a pcl var");
+        }
+      })
+          .Case<ReturnOp>([&b, &outVars](ReturnOp op) {
+        // We return all the output vars we defined above.
+        b.create<pcl::ReturnOp>(
+            op.getLoc(), (llvm::SmallVector<Value>(outVars.begin(), outVars.end()))
+        );
+      }).Default([](Operation *unknown) {
+        unknown->emitError("unsupported op in PCL lowering: ") << unknown->getName();
+      });
+      if (failed(res)) {
+        return failure();
+      }
+    }
+    return success();
+  }
+
+  FailureOr<func::FuncOp> buildPCLFunc(StructDefOp structDef) {
+    SmallVector<Type> pclInputTypes, pclOutputTypes;
+    auto constrainFunc = structDef.getConstrainFuncOp();
+    auto ctx = structDef.getContext();
+    for (auto arg : constrainFunc.getArguments().drop_front()) {
+      if (!llvm::dyn_cast<FeltType>(arg.getType())) {
```

Comment:
```suggestion
      if (!llvm::isa<FeltType>(arg.getType())) {
```

Same comment about casting